### PR TITLE
[SMTChecker] Fix check that solver is available

### LINKED ISSUE #12998: [SMTChecker] ICE when selecting Z3 solver and the dynamic library is not installed
## Description
The compiler reports an Internal Compiler Error when you try to use Z3 solver with a solc binary that uses dynamically linked Z3 and Z3 is not installed.

## Steps to Reproduce
```
curl https://binaries.soliditylang.org/linux-amd64/solc-linux-amd64-v0.8.13+commit.abaa5c0e --output solc-0.8.13
chmod +x solc-0.8.13
echo "contract C {}" | ./solc-0.8.13 - --model-checker-solvers z3 --model-checker-engine all
```
```
Internal compiler error:
/solidity/libsolidity/formal/CHC.cpp(1016): Throw in function void solidity::frontend::CHC::resetSourceAnalysis()
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Solidity assertion failed
[solidity::util::tag_comment*] = Solidity assertion failed
```

## Environment
- Compiler version: 0.8.7 - 0.8.13
- Operating system: Arch Linux

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes https://github.com/ethereum/solidity/issues/12998

### CODE REVIEW

**Comment by ekpyron** [comment]
Maybe also a changelog entry for this?

**Comment by leonardoalt** [comment]
Added

File: libsolidity/formal/CHC.cpp [hunk 1]
```diff
@@ -60,39 +60,42 @@ using namespace solidity::frontend::smt;
...
-	[[maybe_unused]] map<util::h256, string> const& _smtlib2Responses,
-	[[maybe_unused]] ReadCallback::Callback const& _smtCallback,
+	map<util::h256, string> const& _smtlib2Responses,
+	ReadCallback::Callback const& _smtCallback,
 	ModelCheckerSettings const& _settings,
 	CharStreamProvider const& _charStreamProvider
 ):
+	m_smtlib2Responses(_smtlib2Responses),
+	m_smtCallback(_smtCallback),
 	SMTEncoder(_context, _settings, _errorReporter, _charStreamProvider)
```
Comment (ekpyron) [inline libsolidity/formal/CHC.cpp]:
```suggestion
	SMTEncoder(_context, _settings, _errorReporter, _charStreamProvider),
	m_smtlib2Responses(_smtlib2Responses),
	m_smtCallback(_smtCallback)
```
should fix the build

Comment (ekpyron) [reply libsolidity/formal/CHC.cpp]: Or at least half of it... may also have some forward-declaration issues apart from the ordering here...

**Review by ekpyron — APPROVED** [review]

**Comment by ekpyron** [comment]
Damn... haven't seen that this was set to auto-merge... might have asked to squash otherwise...

**Pull request merged**